fastapi_users/db/ormar.py

```diff
@@ -0,0 +1,128 @@
+
+
+class OrmarUserDatabase(BaseUserDatabase[UD]):
+    """
+    Database adapter for ormar.
+
+    :param user_db_model: Pydantic model of a DB representation of a user.
+    :param model: ormar ORM model.
+    :param oauth_account_model: Optional ormar ORM model of a OAuth account.
+    """
+
+    model: Type[OrmarBaseUserModel]
+    oauth_account_model: Optional[Type[OrmarBaseOAuthAccountModel]]
+
+    def __init__(
+        self,
+        user_db_model: Type[UD],
+        model: Type[OrmarBaseUserModel],
+        oauth_account_model: Optional[Type[OrmarBaseOAuthAccountModel]] = None,
+    ):
+        super().__init__(user_db_model)
+        self.model = model
+        self.oauth_account_model = oauth_account_model
+
+    async def get_db_user(self, **kwargs: Any) -> Optional[ormar.Model]:
+        try:
+            query = self.model.objects.filter(**kwargs)
+            if self.oauth_account_model is not None:
+                # in ormar you can select from reverse too
+                # only one query is run - faster execution
+                query = query.select_related("oauth_accounts").order_by(
+                    "oauth_accounts__created_date"
+                )
+            return await query.get()
+        except NoMatch:
+            return None
+
+    async def get_user(self, **kwargs: Any) -> Optional[UD]:
+        user = await self.get_db_user(**kwargs)
+        return self.user_db_model(**user.dict()) if user else None
+
+    async def create_oauth_models(
+        self, model: ormar.Model, oauth_accounts: List[BaseOAuthAccount]
+    ):
+        # here we hardcore the name of the relation despite
+        # that user can give a different one, yet it's the same as
+        # in tortoise backend
+        await self.oauth_account_model.objects.bulk_create(
+            [
+                self.oauth_account_model(user=model, **oacc.dict())
+                for oacc in oauth_accounts
+            ]
+        )
+
+    async def get(self, id: UUID4) -> Optional[UD]:
+        return await self.get_user(id=id)
+
+    async def get_by_email(self, email: str) -> Optional[UD]:
+        return await self.get_user(email__iexact=email)
+
+    async def get_by_oauth_account(self, oauth: str, account_id: str) -> Optional[UD]:
+        return await self.get_user(
+            oauth_accounts__oauth_name=oauth, oauth_accounts__account_id=account_id
+        )
+
+    async def create(self, user: UD) -> UD:
+        oauth_accounts = getattr(user, "oauth_accounts", [])
+        model = await self.model(**user.dict(exclude={"oauth_accounts"})).save()
+        if oauth_accounts and self.oauth_account_model:
+            await self.create_oauth_models(model=model, oauth_accounts=oauth_accounts)
+        return user
+
+    async def update(self, user: UD) -> UD:
+        oauth_accounts = getattr(user, "oauth_accounts", [])
+        model = await self.get_db_user(id=user.id)
+        # have no idea why other backends does not check if user exists?
```

Comment:
What happens in other adapters is that we directly make a request with the ORM. So, if the user does not exist, the ORM exception is directly raised. 

What I propose is that you don't swallow the exception in `get_user_db`, but rather in `get_user`:

```py

    async def get_db_user(self, **kwargs: Any) -> Optional[ormar.Model]:
        query = self.model.objects.filter(**kwargs)
        if self.oauth_account_model is not None:
            # in ormar you can select from reverse too
            # only one query is run - faster execution
            query = query.select_related("oauth_accounts").order_by(
                "oauth_accounts__created_date"
            )
        return await query.get()

    async def get_user(self, **kwargs: Any) -> Optional[UD]:
        try:
            user = await self.get_db_user(**kwargs)
            return self.user_db_model(**user.dict())
        except NoMatch:
            return None
```

This way, you won't have to check it here.

Reply:
That is not true for all adapters.
Simply add this part for sqlalchemy backend tests:

```python
    # Update
    user_db.is_superuser = False # change to False
    await sqlalchemy_user_db.update(user_db)

    # NEW - fail update not fails
    user_db.is_superuser = True
    old_id = user_db.id
    user_db.id = uuid.uuid4()  # new id that does not exist
    await sqlalchemy_user_db.update(user_db)  # should fail but runs query that changes nothing
    user_db.id = old_id 
    # END OF NEW   
    
    # since we not updated anything the old one fails
    # Get by id
    id_user = await sqlalchemy_user_db.get(user.id)
    assert id_user is not None
    assert id_user.id == user_db.id
    assert id_user.is_superuser is True # fails here as there was no update
```

Didn't check the other adapters to be honest.

Reply:
Hmm, maybe there should be a check and a exception in the `update` method, indeed. But this is out of the scope of this PR.

Yes, I confirm that I expect `None` to be returned by `get`, `get_by_email` and `get_by_oauth_account` method. However, you can very well raise it from your private method `get_db_user` ; to ensure an exception is raised if we try to update an unknown user.

As for reinstantiating back the user from the DB response, my assumption is that Pydantic model and DB model should be equivalent. So database should not do anything but writing the data in the Pydantic model.

Reply:
Yes I agree about the exception but if we move the exception check to _get_user it will work fine without additional check in update. But in this version (before my pr to @paolodina fork) `ormar` did not refresh the fields from db, just returned passed user form pydantic (tortoise is doing the same)

What if you have a nullable column that have a server_default setting (possible in `ormar`, in some other backends inqluding `sqlalchemy` too). For example updated_date that should use the sql server datetime and not the python app one.

This field is null when passed from fastapi-users (and can be as it's nullable, no default value in pydantic too) and will be populated only during the query. Or another example is just onupdate "trigger" that always updates the update date but when the passed to db adapter pydantic model is null or the previous one. Without refetch/parsing the db model the returned original model would be outdated.

in example new version for ormar is (to be merged from pr):
```python
    async def create(self, user: UD) -> UD:
        oauth_accounts = getattr(user, "oauth_accounts", [])
        model = await self.model(**user.dict(exclude={"oauth_accounts"})).save()
        if oauth_accounts and self.oauth_account_model:
            await self._create_oauth_models(model=model, oauth_accounts=oauth_accounts)
        user = await self._get_user(id=user.id)  # <- refresh the user from db - make sure it's all updated
        return user
``` 

and was similiar to tortoise (in my opinion it's wrong):
```python
    async def create(self, user: UD) -> UD: # <- user to save is passed
        user_dict = user.dict()
        oauth_accounts = user_dict.pop("oauth_accounts", None)

        model = self.model(**user_dict)  # < - new model is created and saved
        await model.save()

        if oauth_accounts and self.oauth_account_model:
            oauth_account_objects = []
            for oauth_account in oauth_accounts:
                oauth_account_objects.append(
                    self.oauth_account_model(user=model, **oauth_account) # <- same model is passed to oauths
                )
            await self.oauth_account_model.bulk_create(oauth_account_objects)

        return user # < - but only original user that was passed is returned - that does not have to be the same as model and model is never parsed or refetched to user that is returned

```